File: app/src/main/java/io/lerk/lrkFM/activities/file/ArchiveParentFinder.java
package io.lerk.lrkFM.activities.file;

import java.io.File;

import io.lerk.lrkFM.entities.FMArchive;
import io.lerk.lrkFM.entities.FMFile;

/**
 * @author Lukas Fülling (lukas@k40s.net)
 */
public class ArchiveParentFinder {
    private String path;
    private boolean archive;
    private FMArchive archiveFile;

    public ArchiveParentFinder(String path) {
        this.path = path;
    }

    public boolean isArchive() {
        return archive;
    }

    public FMArchive getArchiveFile() {
        return archiveFile;
    }

    public ArchiveParentFinder invoke() {
        archive = false;
        archiveFile = null;
        String tPath = path;
        while (!"/".equals(tPath) && tPath != null) {
            FMFile f = new FMFile(new File(tPath));
            if (f.isArchive()) {
                archive = true;
                archiveFile = new FMArchive(new File(tPath));
            }
            tPath = new File(tPath).getParent();
            if(archive || tPath == null || tPath.isEmpty()) {
                break;
            }
        }
        return this;
    }
}

File: app/src/main/java/io/lerk/lrkFM/activities/file/ArchiveParentFinder.kt
package io.lerk.lrkFM.activities.file

import io.lerk.lrkFM.entities.FMArchive
import io.lerk.lrkFM.entities.FMFile
import io.lerk.lrkFM.exceptions.BlockingStuffOnMainThreadException
import java.io.File

/**
 * Finds parent archives.
 *
 * @author Lukas Fülling (lukas@k40s.net)
 */
class ArchiveParentFinder(private val path: String?) {
    var isArchive = false
        private set
    var archiveFile: FMArchive? = null
        private set

    @Throws(BlockingStuffOnMainThreadException::class)
    operator fun invoke(): ArchiveParentFinder {
        isArchive = false
        archiveFile = null
        var tPath = path
        while ("/" != tPath && tPath != null) {
            val f = FMFile(File(tPath))
            if (f.isArchive) {
                isArchive = true
                archiveFile = FMArchive(File(tPath))
            }
            tPath = File(tPath).parent
            if (isArchive || tPath == null || tPath.isEmpty()) {
                break
            }
        }
        return this
    }
}
